snuba/migrations/snuba_migrations/metrics/0007_granularity_10s.py

```diff
@@ -0,0 +1,97 @@
+        get_forward_view_migration_local(
+            source_table_name="metrics_buckets_local",
+            table_name="metrics_sets_local",
+            mv_name="metrics_sets_mv_local",
+            aggregation_col_schema=[
+                Column("value", AggregateFunction("uniqCombined64", [UInt(64)])),
+            ],
+            aggregation_states="uniqCombined64State(arrayJoin(set_values)) as value",
+            granularity=granularity,
+        ),
+        get_forward_view_migration_local(
+            source_table_name="metrics_counters_buckets_local",
+            table_name="metrics_counters_local",
+            mv_name="metrics_counters_mv_local",
+            aggregation_col_schema=[
+                Column("value", AggregateFunction("sum", [Float(64)])),
+            ],
+            aggregation_states="sumState(value) as value",
+            granularity=granularity,
+        ),
+        get_forward_view_migration_local(
+            source_table_name="metrics_distributions_buckets_local",
+            table_name="metrics_distributions_local",
+            mv_name="metrics_distributions_mv_local",
+            aggregation_col_schema=[
+                Column(
+                    "percentiles",
+                    AggregateFunction(
+                        "quantiles(0.5, 0.75, 0.9, 0.95, 0.99)", [Float(64)]
+                    ),
+                ),
+                Column("min", AggregateFunction("min", [Float(64)])),
+                Column("max", AggregateFunction("max", [Float(64)])),
+                Column("avg", AggregateFunction("avg", [Float(64)])),
+                Column("sum", AggregateFunction("sum", [Float(64)])),
+                Column("count", AggregateFunction("count", [Float(64)])),
+            ],
+            aggregation_states=(
+                "quantilesState(0.5, 0.75, 0.9, 0.95, 0.99)((arrayJoin(values) AS values_rows)) as percentiles, "
+                "minState(values_rows) as min, "
+                "maxState(values_rows) as max, "
+                "avgState(values_rows) as avg, "
+                "sumState(values_rows) as sum, "
+                "countState(values_rows) as count"
+            ),
+            granularity=granularity,
+        ),
+    ]
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    """
+    Re-create materialized views for metrics with a 10 second granularity
+    instead of 60 seconds.
+
+    The forward migration keeps the data rounded to start of minute, as there
+    is no way of recovering granularity for existing entries.
+
+    The backward migration could theoretically round existing entries to start of minute,
+    but that is omitted for the sake of simplicity.
+    """
+
+    blocking = False
+
+    def forwards_local(self) -> Sequence[operations.SqlOperation]:
+        return drop_views() + recreate_views(granularity=10)
+
+    def backwards_local(self) -> Sequence[operations.SqlOperation]:
+        return drop_views() + recreate_views(granularity=60)
```

Comment:
Why are you deleting the 60 seconds materialized views? I thought we wanted both of them.

Reply:
I did not want to get into the business of handling multiple storages for reading. Would keeping both of them mean writing a `QueryStorageSelector` for metrics?

Reply:
We will definitively need to have multiple granularities to deal with. I don't think we will ever be able to get away with reasonable query performance. It will be unreasonable to run a month long query over 10 seconds buckets. 
We are likely to have at least 10 seconds, 1 hour, 1 day. Then maybe more. Not sure about the 1 minute aggregation.

Now since they are all in the same table we have the choice of having multiple "virtual storages" on top of the same table or having only one storage and deal with the granularity in the query processing phase. But this does not change the problem that multiple granularities have to exist. 

Since the format of different granularities here does not change (it does on sessions and outcomes) we could even make it simpler and have that done at Entity level where the `granularity` field is still visible.

Reply:
I think we can keep it simple at first and Have a query processor in the Entity that picks up the granularity field from the query and applies the proper condition to the entity itself. No need for multiple storages. Still the application needs to deal with multiple granularities.

Reply:
@jjbayer, do you need further assistance with this PR ?

Reply:
@fpacifici Thanks, I think I might be able to write the query processor as you suggested, depending on how desperately we need this change. @jan-auer what do you think?

Reply:
So multiple granularities are a must to run this data model in production. 
If you want to have only one for now, I am ok with that, but we are never going to put production load on this without having the hourly and daily aggregation, so it will have to be done anyway. Up to you.

If you build the processor. Please do that as an entity processor not as a storage selector. There you do not need to find patterns in the query. You will instead have access to the granularity field in the query and can process that.

Reply:
So I assume the way you plan to run your queries is generally something like:

```
SELECT bucketed_time
GROUP BY bucketed_time
GRANULARITY something like 10 - 3600 - etc.
```

Is that correct? 
If yes I would expect you plan to use multiple types of granularities always multiple of the minimum one that is 10 seconds. Is that correct?

Then do you ever need the bare timestamp column ?